src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.cpp

```diff
@@ -0,0 +1,224 @@
+#include "NumericalAlgorithms/DiscontinuousGalerkin/Formulation.hpp"
+#include "NumericalAlgorithms/DiscontinuousGalerkin/NormalDotFlux.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"  // IWYU pragma: keep
+
+/// \cond
+namespace ScalarWave::BoundaryCorrections {
+template <size_t Dim>
+UpwindPenalty<Dim>::UpwindPenalty(CkMigrateMessage* msg) noexcept
+    : BoundaryCorrection<Dim>(msg) {}
+
+template <size_t Dim>
+std::unique_ptr<BoundaryCorrection<Dim>> UpwindPenalty<Dim>::get_clone()
+    const noexcept {
+  return std::make_unique<UpwindPenalty>(*this);
+}
+
+template <size_t Dim>
+void UpwindPenalty<Dim>::pup(PUP::er& p) {
+  BoundaryCorrection<Dim>::pup(p);
+}
+
+template <size_t Dim>
+double UpwindPenalty<Dim>::dg_package_data(
+    const gsl::not_null<Scalar<DataVector>*> packaged_char_speed_v_psi,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        packaged_char_speed_v_zero,
+    const gsl::not_null<Scalar<DataVector>*> packaged_char_speed_v_plus,
+    const gsl::not_null<Scalar<DataVector>*> packaged_char_speed_v_minus,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        packaged_char_speed_n_times_v_plus,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        packaged_char_speed_n_times_v_minus,
+    const gsl::not_null<Scalar<DataVector>*> packaged_char_speed_gamma2_v_psi,
+    const gsl::not_null<tnsr::i<DataVector, 3, Frame::Inertial>*>
+        packaged_char_speeds,
+
+    const Scalar<DataVector>& pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& phi,
+    const Scalar<DataVector>& psi,
+
+    const Scalar<DataVector>& constraint_gamma2,
+
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& normal_covector,
+    const std::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+    /*mesh_velocity*/,
+    const std::optional<Scalar<DataVector>>& normal_dot_mesh_velocity)
+    const noexcept {
+  if (normal_dot_mesh_velocity.has_value()) {
+    get<0>(*packaged_char_speeds) = -get(*normal_dot_mesh_velocity);
+    get<1>(*packaged_char_speeds) = 1.0 - get(*normal_dot_mesh_velocity);
+    get<2>(*packaged_char_speeds) = -1.0 - get(*normal_dot_mesh_velocity);
+  } else {
+    get<0>(*packaged_char_speeds) = 0.0;
+    get<1>(*packaged_char_speeds) = 1.0;
+    get<2>(*packaged_char_speeds) = -1.0;
+  }
+
+  // Computes the contribution to the boundary correction from one side of the
+  // interface.
+  //
+  // Note: when UpwindPenalty::dg_boundary_terms() is called, an Element passes
+  // in its own packaged data to fill the interior fields, and its neighbor's
+  // packaged data to fill the exterior fields. This introduces a sign flip for
+  // each normal used in computing the exterior fields.
+  // Use v_psi allocation as n^i Phi_i
+  dot_product(packaged_char_speed_v_psi, normal_covector, phi);
+  get(*packaged_char_speed_gamma2_v_psi) = get(constraint_gamma2) * get(psi);
+  {
```

Comment:
in code comment state that the scope is to limit the region in which `packaged_char_speed_v_psi` and `packaged_char_speed_gamma2_v_psi` are borrowed as buffers.
Also, I think this would be less confusing if the temporary use as a buffer was changed to:
```
auto& normal_dot_phi = get(*packaged_char_speed_v_psi);
dot_product(make_not_null(&normal_dot_phi), normal_covector, phi));
```
(i.e. just consistently treat manipulations associated with a particular temporary identity for the buffer to avoid confusion associated with the variable names). If you choose to do this, maybe consider similar for `packaged_char_speed_gamma2_v_psi`.

Reply:
I agree with Jordan's comment and it doesn't look addressed to me. I'd like to see
- the comment on line 79 move one line down (to below line 80), and say "In this scope, use v_psi allocation as n^i Phi_i"
- then reorder the `dot_product` and the alias creation lines below, so the scope starts with creating the `normal_dot_phi` alias.

Reply:
point 2: I think the reorder is a bad decision because then `normal_dot_phi` can't be const _and_ I'd need to do a `make_not_null`

Reply:
okay, that's a fair counter-argument 👍